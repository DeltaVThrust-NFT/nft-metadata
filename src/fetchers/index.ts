import { fetchUnderlyingContractData } from './defaultFetcher'
import { getAddress } from '@ethersproject/address'

export function fetcherLookup(contractAddress: string) {
  const safeAddress = getAddress(contractAddress)
  switch (safeAddress) {
    default:
      return fetchUnderlyingContractData
  }
}
