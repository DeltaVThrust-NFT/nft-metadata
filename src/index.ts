export * from './constants'
export * from './metadata'
export * from './agent'
export * from './uri'
