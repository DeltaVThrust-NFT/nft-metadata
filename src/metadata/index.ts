export * from './fetcher'
export * from './hashmasks'
export * from './loot'
export * from './normaliser'
export * from './zora'
