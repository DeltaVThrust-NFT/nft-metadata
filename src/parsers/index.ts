import { getAddress } from '@ethersproject/address'
import { parseGenericMetadata } from './openseaMetadataParser'
import {
  ART_BLOCKS_CURATED_TOKEN_ADDRESS,
  ART_BLOCKS_TOKEN_ADDRESS,
  HASHMASKS_TOKEN_ADDRESS,
  MAKERSPLACE_TOKEN_ADDRESS,
  ZORA_TOKEN_ADDRESS,
} from '../constants'
import { parseZoraMetadata } from './zoraMetadataParser'
import { parseMakersplaceMetadata } from './makersplaceMetadataParser'
import { parseHashmasksMetadata } from './hashmasksMetadataParser'
import { parseArtblocksMetadata } from './artblocksMetadataParser'

export function parserLookup(contractAddress: string) {
  const safeAddress = getAddress(contractAddress)
  switch (safeAddress) {
    case ART_BLOCKS_TOKEN_ADDRESS:
      return parseArtblocksMetadata
    case ART_BLOCKS_CURATED_TOKEN_ADDRESS:
      return parseArtblocksMetadata
    case HASHMASKS_TOKEN_ADDRESS:
      return parseHashmasksMetadata
    case MAKERSPLACE_TOKEN_ADDRESS:
      return parseMakersplaceMetadata
    case ZORA_TOKEN_ADDRESS:
      return parseZoraMetadata
    default:
      return parseGenericMetadata
  }
}
