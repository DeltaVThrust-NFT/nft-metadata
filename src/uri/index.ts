export { getAlternateContractCall } from './contract'
export {
  fetchMimeType,
  fetchURI,
  fetchWithRetries,
  fetchWithTimeout,
  parseDataUri,
} from './fetch'
export { isIPFS, getIPFSUrl, getPrivateGateway } from './ipfs'
export { getStaticURI, getURIData } from './static'
