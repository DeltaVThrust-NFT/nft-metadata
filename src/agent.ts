import { JsonRpcProvider } from '@ethersproject/providers'
import { Networkish } from '@ethersproject/networks'
import {
  fetchMimeType,
  fetchURI,
  getAlternateContractCall,
  getIPFSUrl,
  getPrivateGateway,
  getStaticURI,
  getURIData,
  IPFS_IO_GATEWAY,
} from './uri'
import { getAddress } from '@ethersproject/address'
import { Erc721Factory } from '@zoralabs/core/dist/typechain'
import { fetchOnChainData, normaliseURIData } from './metadata'

type AgentOptions = {
  network: Networkish
  networkUrl: string
  ipfsGatewayUrl?: string
  timeout?: number
}

export interface NftMetadata {
  tokenId: string
  tokenAddress: string
  metadata: any
  tokenURI: string
  tokenURL: string
  tokenURLMimeType: string
  name?: string
  description?: string
  contentURL?: string
  contentURLMimeType?: string
  imageURL?: string
  imageURLMimeType?: string
  externalLink?: string
  attributes?: Record<string, any>[]
}

export class Agent {
  timeout: number
  ipfsGatewayUrl: string
  provider: JsonRpcProvider

  constructor(options: AgentOptions) {
    this.provider = new JsonRpcProvider(options.networkUrl, options.network)
    this.ipfsGatewayUrl = options.ipfsGatewayUrl || IPFS_IO_GATEWAY
    this.timeout = options.timeout || 10000
  }

  public async fetchTokenURI(tokenAddress: string, tokenId: string) {
    const staticURI = getStaticURI(tokenAddress, tokenId)
    if (staticURI) {
      return staticURI
    }
    const alternateMethod = getAlternateContractCall(
      tokenAddress,
      tokenId,
      this.provider,
    )
    if (alternateMethod) {
      return alternateMethod
    }
    const contract = Erc721Factory.connect(tokenAddress, this.provider)
    try {
      return contract.tokenURI(tokenId)
    } catch (e) {
      return
    }
  }

  public async fetchURIData(
    tokenAddress: string,
    tokenId: string,
    tokenURI: string,
    ipfsGateway: string,
  ) {
    const alternateMethod = getURIData(tokenAddress, tokenId)
    if (alternateMethod) {
      return alternateMethod
    }
    const resp = await fetchURI(
      tokenURI,
      { timeout: this.timeout },
      ipfsGateway,
    )
    if (!resp) {
      throw new Error(`Failed to fetch uri data for token from: ${tokenURI}`)
    }
    return resp
  }

  private async parseURIData(
    tokenAddress: string,
    tokenId: string,
    tokenURI: string,
    uriData: any,
    ipfsGateway: string,
  ) {
    const onChainData = await fetchOnChainData(
      tokenAddress,
      tokenId,
      this.provider,
    )
    const meta = normaliseURIData(tokenAddress, {
      ...uriData,
      ...onChainData,
      ...(uriData.mimeType && {
        contentURLMimeType: uriData.mimeType,
      }),
    })

    if (meta.image) {
      meta.imageURL = getIPFSUrl(meta.image, ipfsGateway)
    }

    if (meta.animation_url) {
      meta.contentURL = getIPFSUrl(meta.animation_url, ipfsGateway)
    }

    if (!meta.contentURL && meta.imageURL) {
      meta.contentURL = meta.imageURL
    }

    if (meta.contentURL && !meta.contentURLMimeType) {
      meta.contentURLMimeType = await fetchMimeType(meta.contentURL, {
        timeout: this.timeout,
      })
    }

    if (meta.imageURL && !meta.imageURLMimeType) {
      meta.imageURLMimeType = await fetchMimeType(meta.imageURL, {
        timeout: this.timeout,
      })
    }

    const {
      name,
      description,
      attributes,
      external_url: externalURL,
      imageURL,
      imageURLMimeType,
      contentURL,
      contentURLMimeType,
    } = meta

    return {
      tokenURL: getIPFSUrl(tokenURI, ipfsGateway),
      tokenURLMimeType: 'application/json',
      ...(name && { name }),
      ...(description && { description }),
      ...(imageURL && { imageURL }),
      ...(imageURLMimeType && { imageURLMimeType }),
      ...(contentURL && { contentURL }),
      ...(contentURLMimeType && { contentURLMimeType }),
      ...(attributes && { attributes }),
      ...(externalURL && { externalURL }),
    }
  }

  public async fetchMetadata(
    rawAddress: string,
    tokenId: string,
  ): Promise<NftMetadata> {
    const tokenAddress = getAddress(rawAddress)
    const tokenURI = await this.fetchTokenURI(tokenAddress, tokenId)
    if (!tokenURI) {
      throw new Error(
        `Failed to get tokenURI token: ${tokenAddress} is unsupported by @zoralabs/nft-metadata`,
      )
    }
    console.log('fetched uri ', { tokenURI })
    const ipfsGateway = getPrivateGateway(tokenAddress) || this.ipfsGatewayUrl
    const URIData = await this.fetchURIData(
      tokenAddress,
      tokenId,
      tokenURI,
      ipfsGateway,
    )
    console.log('fetched uri data: ', { URIData })
    const metadata = await this.parseURIData(
      tokenAddress,
      tokenId,
      tokenURI,
      URIData,
      ipfsGateway,
    )
    console.log('parsed metada: ', { metadata })

    return {
      tokenId,
      tokenAddress,
      metadata: URIData,
      tokenURI,
      ...metadata,
    }
  }
}
