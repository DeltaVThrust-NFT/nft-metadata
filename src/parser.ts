import { JsonRpcProvider } from '@ethersproject/providers'
import { getAddress } from '@ethersproject/address'
import { fetcherLookup } from './fetchers'
import { parserLookup } from './parsers'

export interface NftMetadata {
  metadata: any

  name?: string
  description?: string

  ownerAddress: string

  tokenURI: string
  contentURI: string

  externalLink?: string
  attributes?: Record<string, any>[]
}

export class Parser {
  constructor(private readonly provider: JsonRpcProvider) {}

  public async fetchUnderlyingContractData(
    contractAddress: string,
    tokenId: string,
  ) {
    const fetcher = fetcherLookup(contractAddress)
    return fetcher(this.provider, contractAddress, tokenId)
  }

  public async parseTokenMetadata(
    contractAddress: string,
    tokenId: string,
    tokenURI: string,
  ) {
    const parser = parserLookup(contractAddress)
    return parser(this.provider, contractAddress, tokenId, tokenURI)
  }

  public async fetchAndParseTokenMeta(
    tokenAddress: string,
    tokenId: string,
  ): Promise<NftMetadata> {
    const safeAddress = getAddress(tokenAddress)

    const { tokenURI, ownerAddress } = await this.fetchUnderlyingContractData(
      safeAddress,
      tokenId,
    )

    const parsedMetadata = await this.parseTokenMetadata(
      safeAddress,
      tokenId,
      tokenURI,
    )

    return {
      tokenURI,
      ownerAddress,
      ...parsedMetadata,
    }
  }
}
