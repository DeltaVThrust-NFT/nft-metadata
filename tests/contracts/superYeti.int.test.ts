import { MetadataAgent, SUPER_YETI_TOKEN_ADDRESS } from '../../src'
import METADATA_STUB from '../mock-reponses/contracts/superYeti/100.json'
import { testProvider } from '../setupProvider'
import { isAddress } from '@ethersproject/address'

const SUPER_YET_CRITERIA = {
  input: {
    tokenId: '100',
    tokenAddress: SUPER_YETI_TOKEN_ADDRESS,
  },
  output: {
    metadata: METADATA_STUB,
    name: METADATA_STUB.name,
    description: METADATA_STUB.description,
    tokenURL:
      'https://defra.systems/metadata/QmXFtqihiEDP5sJwME5dNB3NnYk6LeiepbD4RPP1XES6Ys/asset/100',
    tokenURLMimeType: 'application/json; charset=utf-8',
    contentURL: METADATA_STUB.image,
    contentURLMimeType: 'image/jpeg',
    attributes: METADATA_STUB.attributes,
  },
}

describe('Super Yeti ERC721', () => {
  const parser = new MetadataAgent(testProvider)

  beforeEach(() => {
    jest.setTimeout(10000)
  })

  it(`should be able to fetch and parse metadata for token id: ${SUPER_YET_CRITERIA.input.tokenId}`, async () => {
    const { ownerAddress, ...meta } = await parser.fetchAndParseTokenMeta(
      SUPER_YET_CRITERIA.input.tokenAddress,
      SUPER_YET_CRITERIA.input.tokenId,
    )
    expect(meta).toStrictEqual(SUPER_YET_CRITERIA.output)
    expect(isAddress(ownerAddress)).toBeTruthy()
  })
})
