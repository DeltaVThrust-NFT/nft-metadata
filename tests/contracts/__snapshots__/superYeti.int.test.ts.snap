// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Super Yeti ERC721 should be able to fetch and parse metadata for token id: 100 on network: 1 1`] = `
Object {
  "attributes": Array [
    Object {
      "trait_type": "Background Color",
      "value": "Yellow",
    },
    Object {
      "trait_type": "Fur Color",
      "value": "Red Black",
    },
    Object {
      "trait_type": "Eye Color",
      "value": " Black",
    },
    Object {
      "trait_type": "Face",
      "value": "Werewolf",
    },
    Object {
      "value": "DSLR",
    },
    Object {
      "value": "Top Hat Gold",
    },
    Object {
      "value": "Ski Goggles",
    },
    Object {
      "value": "Ski Boots",
    },
    Object {
      "trait_type": "Rarity",
      "value": "Prime",
    },
    Object {
      "trait_type": "Track",
      "value": "Photographer",
    },
  ],
  "contentURL": "https://defra.mypinata.cloud/ipfs/Qmbhrr2skR8au5Wh8HtJhW4PYL7QjEUUnXC2r6BTwZVuYB",
  "contentURLMimeType": "image/jpeg",
  "description": "SuperYeti is a 3D collection of randomized characters created through a standard ERC-721 contract and backed by IPFS file storage and registered on the Ethereum Blockchain. Future implementation of a multiplayer, play-to-earn game where SuperYetis own land, breed and compete with one another to protect the Yetiverse. You can learn more at www.superyeti.co",
  "metadata": Object {
    "attributes": Array [
      Object {
        "trait_type": "Background Color",
        "value": "Yellow",
      },
      Object {
        "trait_type": "Fur Color",
        "value": "Red Black",
      },
      Object {
        "trait_type": "Eye Color",
        "value": " Black",
      },
      Object {
        "trait_type": "Face",
        "value": "Werewolf",
      },
      Object {
        "value": "DSLR",
      },
      Object {
        "value": "Top Hat Gold",
      },
      Object {
        "value": "Ski Goggles",
      },
      Object {
        "value": "Ski Boots",
      },
      Object {
        "trait_type": "Rarity",
        "value": "Prime",
      },
      Object {
        "trait_type": "Track",
        "value": "Photographer",
      },
    ],
    "description": "SuperYeti is a 3D collection of randomized characters created through a standard ERC-721 contract and backed by IPFS file storage and registered on the Ethereum Blockchain. Future implementation of a multiplayer, play-to-earn game where SuperYetis own land, breed and compete with one another to protect the Yetiverse. You can learn more at www.superyeti.co",
    "image": "https://defra.mypinata.cloud/ipfs/Qmbhrr2skR8au5Wh8HtJhW4PYL7QjEUUnXC2r6BTwZVuYB",
    "name": "Super Yeti #100",
  },
  "name": "Super Yeti #100",
  "networkId": 1,
  "tokenAddress": "0x3F0785095A660fEe131eEbcD5aa243e529C21786",
  "tokenId": "100",
  "tokenURI": "https://defra.systems/metadata/QmXFtqihiEDP5sJwME5dNB3NnYk6LeiepbD4RPP1XES6Ys/asset/100",
  "tokenURL": "https://defra.systems/metadata/QmXFtqihiEDP5sJwME5dNB3NnYk6LeiepbD4RPP1XES6Ys/asset/100",
  "tokenURLMimeType": "application/json; charset=utf-8",
}
`;
